HPCC-20631 Set session id context

mayx opened the pull request:
- Set the session id context of an ssl context to be "hpccsystems", a
  value unique to our applicaiton.

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

mayx commented:
@mckellyln please review.

HPCCSmoketest commented:
This PR (11764) is in the 1/2 position of the queue. Estimated start time is after ~0.5 hour(s)

HPCCSmoketest commented:
2/3. Process of PR-11764, label: HPCC-20631-VerifyCert starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: dd2b1e88f0a6de425979b0f170d2cb46eaf50cc0
Build: success
Install hpccsystems-platform-community_7.0.0-rc2.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   26 sec (00:00:26) |  187 sec (00:03:07) |   95 sec (00:01:35) |   19 sec (00:00:19) |   42 sec (00:00:42) | 1461 sec (00:24:21) |   18 sec (00:00:18) | 1848 sec (00:30:48) |

mckellyln commented on system/security/securesocket/securesocket.cpp:
```diff
@@ -896,6 +900,9 @@ class CSecureSocketContext : implements ISecureSocketContext, public CInterface
             throw MakeStringException(-1, "ctx can't be created");
         }
+
+        setSessionIdContext();
```
I think setting the ctx session is meaningful only on the server side ?
If that is true, should we do this only for server, ie:
```
        if(sockettype == ServerSocket)
            setSessionIdContext();
```

mckellyln commented:
@mayx a few questions:
1). does it make sense to set on client side ?
2). would it make sense to have different ids for different components or use cases ?
3). what is the point of setting the id, are you planning on adding session re-use/cache support ?

mayx commented:
1) it's only needed on the server side
2) Not for our use case
3) The purpose is to fix an error that happens with SSL_accept when client certificate verification is enabled and the client sends in a certificate. Client certificate verification is usually not enabled.

mckellyln commented:
@mayx ok, thanks - so should it just be done before SSL_accept() or only if sockettype == ServerSocket ?
Other than that - looks good.

mayx commented:
Good point. The man page states it's only "useful" for server side and it doesn't seem to affect client side functioning. But to stay on the safe side, I'll add the check of sockettype.

mayx committed:
HPCC-20631 Set session id context

- Set the session id context of an ssl context to be "hpccsystems", a
  value unique to our applicaiton.

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

mayx commented:
@mckellyln change uploaded please review again.

HPCCSmoketest commented:
This PR (11764) is in the 1/2 position of the queue. Estimated start time is after ~0.5 hour(s)

HPCCSmoketest commented:
2/3. Process of PR-11764, label: HPCC-20631-VerifyCert starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 4a1f1dcd800df42b3240673557f3e8cf290d9c25
Build: success
Install hpccsystems-platform-community_7.0.0-rc2.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   21 sec (00:00:21) |  190 sec (00:03:10) |   95 sec (00:01:35) |   19 sec (00:00:19) |   42 sec (00:00:42) | 1670 sec (00:27:50) |   18 sec (00:00:18) | 2055 sec (00:34:15) |

mckellyln commented:
@mayx looks good.

mckellyln reviewed (approved):
@mayx approved
